### PR TITLE
[MXNET-853] Fix for smooth_l1 operator scalar default value

### PR DESCRIPTION
**Pull request by samskalicky**
## Description ##
Fix for issue #7533 Call symbol/nd.smooth_l1 without scalar causing uncaught exception

changed smooth_l1 operator implementation to not use helper macros since
they do not provide enough support for checking for arguments and
setting custom default values

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
changed smooth_l1 operator implementation to not use helper macros since
they do not provide enough support for checking for arguments and
setting custom default values

### CODE REVIEW

**Comment by zhreshold** [comment]
It should properly catch the exception anyway, rather than assigning every single operator a default value

**Comment by samskalicky** [comment]
@haojin2 linting fixed
@zhreshold Is catching the exception what we want to do? What should we do after catching the exception? Set the default value? If we dont set something, its just going to error out somewhere further down the line. It seems to me that this is actually a larger and more in-depth problem than it initially appears so ive filed another issue to track that specifically #12286. But fixing that is going require a little more planning. What is the minimum acceptable behavior for this operator to close this issue?

File: src/operator/tensor/elemwise_binary_scalar_op_extended.cc [hunk 1]
```diff
@@ -98,17 +98,37 @@ where :math:`x` is an element of the tensor *lhs* and :math:`\sigma` is the scal
...
+      attrs->parsed = std::stod(attrs->dict["scalar"]);
+    else
+      attrs->parsed = 1.0;
+  })
+.set_attr<nnvm::FInferShape>("FInferShape", ElemwiseShape<1, 1>)
+.set_attr<nnvm::FInferType>("FInferType", ElemwiseType<1, 1>)
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+				[](const NodeAttrs& attrs){
+				  return std::vector<std::pair<int, int> >{{0, 0}};
+				})
```
[the remark this answers is not in the document]
Comment (samskalicky) [reply src/operator/tensor/elemwise_binary_scalar_op_extended.cc]: sorry, new editor not setup properly....

**Comment by zhreshold** [comment]
@samskalicky I didn't mean not sending this PR to fix the problem, but should address the exception catch problem anyways. Sorry about the confusion. #12286 is a perfect attempt. Thanks!

This LGTM now.

**Comment by samskalicky** [comment]
@anirudh2290 can you help review/merge?

File: src/operator/tensor/elemwise_binary_scalar_op_extended.cc [hunk 2]
```diff
@@ -98,17 +98,38 @@ where :math:`x` is an element of the tensor *lhs* and :math:`\sigma` is the scal
...
+.add_argument("scalar", "float", "scalar input")
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, mshadow_op::smooth_l1_loss>)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{ "_backward_smooth_l1" });
 
 MXNET_OPERATOR_REGISTER_BINARY(_backward_smooth_l1)
-.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
-.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<
-  cpu, mshadow_op::smooth_l1_gradient>);
+  .set_attr_parser([](NodeAttrs *attrs) {
+  if (attrs->dict.find("scalar") != attrs->dict.end())
```
Comment (anirudh2290) [inline src/operator/tensor/elemwise_binary_scalar_op_extended.cc]: nit: even one line ifs in MXNet codebase are put inside curly braces.

**Pull request merged**